Organization frame

NicholasNagy opened the pull request:
Fixed   #106 , #117 , #120 .
Additionally, fixed the settings button on the sidenav to route to a settings url instead of the the modify-members url, since this makes more sense.
I also fixed it so that users will only see the appropriate listing of employees depending if they are an organization frame or not.
I modified the components of editing profile to display the UI similar to what we have on account creation.
I also fixed a minor bug in the permissions when creating an account (needed to convert from integer to string in the org comparison).
Fixed warnings in modify-members component that were screaming because the values weren't always instantiated before wanting to be displayed in the html (async issue)

Finished:
- move logout button to bottom of sidenav
- properly display selected sidenav option
- build functionality for which sidenav option is selected
- Move link on prev ALTA button to sidenav ALTA
- Fix height of ALTA square in sidenav
- Properly adjust width and position with top app bar
- Fix all the tests I broke
- Fix lint

makchamp reviewed (commented):
nice work, I would just put the name of the organization somewhere for clarity

Amanda-Kolopa commented:
First off, I love the improvements to the UI! Great work :smile: 

Although, I noticed issues when attempting to log in as an SA:

- After selecting an organization to manage and refreshing the page, the sidenav component displays the options intended for an SA rather than the options for IM.

- After selecting an organization to manage, when I select the Employees tab and refresh the page, I get an error from accessClients saying I'm unauthorized to view this data. (Also I get an error from manage-members, could be a possible hint)

- When I log in, refresh the page, log out, and try to log in again, a get an error for an invalid token (to fix, I have to refresh again)

I also noticed issues when logging in as an IM:

- when I first log in, the sidenav displays the options for an SA. This is because the organization_id is not set in local storage upon login (once I refresh the page, all is good again).

- when I select the Employees tab, I encounter the same problem as before when logged in as an SA

Side note:

- in the Employees section, can we have the "Create" function as a button? I found that it was a bit difficult to locate it at first.

Aside from these issues, the overall structure of the system is looking good!

NicholasNagy commented on client/src/app/services/auth.service.ts:
```diff
@@ -35,20 +38,45 @@ export class AuthService {
+  }
+
+  turnOffOrgMode(): void {
+    if (this.role.getValue() === 'SA') {
```
Could you elaborate?

NicholasNagy commented on client/src/app/services/auth.service.ts:
```diff
@@ -35,20 +38,45 @@ export class AuthService {
+          this.username.next(data.user_name);
+          this.role.next(data.role);
+          this.organizationId.next(data.organization);
+          // TODO: update GET call to return organization's name
```
Done.

NicholasNagy commented on client/src/app/components/sidenav/sidenav.component.ts:
```diff
@@ -10,24 +12,73 @@ import { AuthService } from 'src/app/services/auth.service';
+    this.tokenService.DeleteToken(); // Delete token when user logout
+    this.authService.setLogOut();   // Extra step - sets the sharedUser data to ''
+    this.router.navigate(['login']); // Redirect user to login/register pager
+    // TODO: Check out if we want to delete also the token from the db, in order to regenerate a new one while logging in
```
yeah, this is just so that the token doesn't stay lying around while the user isn't using their pc. Plus, if someone visit the url when they have the token, they will be redirected to the dashboard.

NicholasNagy commented:
> Although, I noticed issues when attempting to log in as an SA:
> 
> * After selecting an organization to manage and refreshing the page, the sidenav component displays the options intended for an SA rather than the options for IM.

Fixed!

> * After selecting an organization to manage, when I select the Employees tab and refresh the page, I get an error from accessClients saying I'm unauthorized to view this data. (Also I get an error from manage-members, could be a possible hint)
> * When I log in, refresh the page, log out, and try to log in again, a get an error for an invalid token (to fix, I have to refresh again)

I cannot reproduce these. Can you pull the new changes and validate that it is still the case?


> I also noticed issues when logging in as an IM:
> 
> * when I first log in, the sidenav displays the options for an SA. This is because the organization_id is not set in local storage upon login (once I refresh the page, all is good again).

Fixed!

> * when I select the Employees tab, I encounter the same problem as before when logged in as an SA

I cannot reproduce this. Can you pull the new changes and validate that it is still the case?

> Side note:
> 
> * in the Employees section, can we have the "Create" function as a button? I found that it was a bit difficult to locate it at first.

It actually was a button... lol, but now it should look better and be easier to find now.

Amanda-Kolopa reviewed (changes requested):
Thank you for fixing the mentioned issues!

I noticed some more issues:

- When logged in as an Inventory Manager and viewing any page other than Dashboard, if I refresh the page, I'm redirected back to Dashboard. 

- When logged in as an Inventory Manager and attempting to edit an employee's role, I can set their role to a System Administrator. This occurs for all the employees except for the logged-in user's listing (next comment).

- In the PR comment you mentioned that you fixed on bug issue #117 but when logged in as an Inventory Manager and viewing the list of employees, I am able to see the logged-in user's data after I refresh the page. (no issues with System Administrator login!) 

Some comments:

- When logged in as a System Administrator and viewing the list of other system administrators in _Manage Members_, I am able to change a user's role to inventory manager or stock-keeper but they will not be affiliated with an organization. This could lead to inconsistency.

- Similar to the point above, a System Administrator is able to alter the role of an Inventory Manager or Stock-Keeper to a System Administrator (perhaps we should remove the functionality of altering roles).

We can further discuss the comments but aside from all this everything is looking great!
Awesome work :smile:

NicholasNagy commented:
> * When logged in as an Inventory Manager and viewing any page other than Dashboard, if I refresh the page, I'm redirected back to Dashboard.

I see this issue, but I think for the requirements of this sprint (epic 1), this can be opened as a bug report for release 2 to handle. I'm really unsure as to where this issue is coming from and Inventory Manager wasn't actually supposed to be covered this release anyway.

> * When logged in as an Inventory Manager and attempting to edit an employee's role, I can set their role to a System Administrator. This occurs for all the employees except for the logged-in user's listing (next comment).

This is a separate user story altogether. Making these modifications would be out of scope for this release.

> * In the PR comment you mentioned that you fixed on bug issue #117 but when logged in as an Inventory Manager and viewing the list of employees, I am able to see the logged-in user's data after I refresh the page. (no issues with System Administrator login!)

I don't observe these results locally and I don't know how you're able to stay on the page if when you refresh, you get redirected to dashboard. Are you sure you are logged in as an Inventory Manager?

> Some comments:
> 
> * When logged in as a System Administrator and viewing the list of other system administrators in _Manage Members_, I am able to change a user's role to inventory manager or stock-keeper but they will not be affiliated with an organization. This could lead to inconsistency.

That's a good point, however, I think this is an unforeseen case that we haven't thought of with Ajay. This will most likely have to be added as a user story for the next release.

> * Similar to the point above, a System Administrator is able to alter the role of an Inventory Manager or Stock-Keeper to a System Administrator (perhaps we should remove the functionality of altering roles).

Well, I think they should have the ability to do so, but this feature was in our requirements. We can discuss this further with Ajay tomorrow.

Amanda-Kolopa reviewed (approved):
After discussion and reviewing the mentioned issues with others, it seems like this PR is good to be merged!
I will make the appropriate bug reports for the mentioned issues and jot down the notes to discuss role assignments with Ajay.
